dandi/dandiapi.py

```diff
@@ -0,0 +1,348 @@
+from contextlib import contextmanager
+
+import requests
+
+from . import get_logger
+from .utils import ensure_datetime
+from .consts import MAX_CHUNK_SIZE
+
+lgr = get_logger()
+
+
+class HTTPError(requests.HTTPError):
+    """An HTTP error occurred.
+
+    Following Girder's recommendation of having its HttpError deprecated,
+    this is just a helper to bring that HTTPError into our space
+    """
+
+    pass
+
+
+# Following class is loosely based on GirderClient, with authentication etc
+# being stripped.
+# TODO: add copyright/license info
+class RESTFullAPIClient(object):
+    """A base class for REST clients"""
+
+    def __init__(self, api_url):
+        self.api_url = api_url
+        self._session = None
+
+    @contextmanager
+    def session(self, session=None):
+        """
+        Use a :class:`requests.Session` object for all outgoing requests.
+        If `session` isn't passed into the context manager
+        then one will be created and yielded. Session objects are useful for enabling
+        persistent HTTP connections as well as partially applying arguments to many
+        requests, such as headers.
+
+        Note: `session` is closed when the context manager exits, regardless of who
+        created it.
+
+        .. code-block:: python
+
+            with client.session() as session:
+                session.headers.update({'User-Agent': 'myapp 1.0'})
+
+                for item in items:
+                    client.downloadItem(item, fh)
+
+        In the above example, each request will be executed with the User-Agent header
+        while reusing the same TCP connection.
+
+        :param session: An existing :class:`requests.Session` object, or None.
+        """
+        self._session = session if session else requests.Session()
+
+        yield self._session
```

Comment:
This should be:

```
try:
    yield self._session
finally:
    self._session.close()
    self._session = None
```

so that the session gets closed on error.

Reply:
indeed needs a fix, thanks!  but it should close only if session was not provided (should not close outside session). Fixed in f367313da0b87f86c1138666b50bf27b34cebab9